state/watcher/txnwatcher.go

```diff
@@ -321,52 +305,193 @@ func (w *TxnWatcher) loop() error {
+	db := w.session.DB(w.jujuDBName)
+	var err error
+	for i := 0; i < 2; i++ {
+		err = w.runCmd(db, bson.D{
+			{"killCursors", "$cmd.aggregate"},
+			{"cursors", []int64{w.cursorId}},
+		}, nil)
+		if isCursorNotFoundError(err) {
+			break
+		} else if err == nil {
+			break
+		}
+		w.logger.Warningf("failed to kill cursor %d: %s", w.cursorId, err.Error())
+	}
+	w.cursorId = 0
+}
+
+func (w *TxnWatcher) init() (bool, error) {
+	db := w.session.DB(w.jujuDBName)
+
+	cs := bson.M{
+		"fullDocument":             "updateLookup",
+		"fullDocumentBeforeChange": "off",
+	}
+	if len(w.resumeToken.Data) > 0 {
+		cs["resumeAfter"] = w.resumeToken
+	}
+
+	match := bson.D{
+		{"$or", []bson.D{
+			{{"operationType", "insert"}},
+			{{"operationType", "update"}},
+			{{"operationType", "replace"}},
+			{{"operationType", "delete"}},
+		}},
+	}
+
+	if len(w.watchCollections) > 0 {
+		match = append(match, bson.DocElem{"ns.coll", bson.D{{"$in", w.watchCollections}}})
+	}
+
+	cmd := bson.D{
+		{"aggregate", 1},
+		{"pipeline", []bson.D{
+			{{"$changeStream", cs}},
+			{{"$match", match}},
+		}},
+		{"cursor", bson.D{{"batchSize", 10}}},
+		{"readConcern", bson.D{{"level", "majority"}}},
+	}
+
+	resp := aggregateResponse{}
+	err := w.runCmd(db, cmd, &resp)
+	if err != nil {
+		return false, errors.Annotate(err, "starting change stream")
+	}
+	w.cursorId = resp.Cursor.Id
+	added, err := w.process(resp.Cursor.FirstBatch)
+	if err != nil {
+		return false, errors.Annotate(err, "processing first change stream batch")
+	}
+	w.resumeToken = resp.Cursor.PostBatchResumeToken
+
+	return added, nil
+}
+
+func (w *TxnWatcher) process(changes []bson.Raw) (bool, error) {
+	added := false
+
+	for i := len(changes) - 1; i >= 0; i-- {
+		changeRaw := changes[i]
+		change := changeStreamDocument{}
+		err := bson.Unmarshal(changeRaw.Data, &change)
+		if err != nil {
+			return added, errors.Trace(err)
+		}
+		if len(change.Id.Data) == 0 {
+			return added, errors.Annotate(FatalChangeStreamError, "missing change id in change")
+		}
+		revno := int64(0)
+		switch change.OperationType {
+		case "insert", "replace":
+			// We read the revno from the committed document as reported by the change stream.
+			revno = change.FullDocument.Revno
+		case "update":
+			// For updates we read the revno from the updated fields so that it isn't newer as
+			// FullDocument on an update is only guaranteed to be the state of the document after
+			// the update, not the state of the document as a result of the update, so the reported
+			// revno could be greater.
+			revno = change.UpdateDescription.UpdatedFields.Revno
+			if revno == 0 && change.TxnNumber == 0 {
+				a := bson.M{}
+				err := bson.Unmarshal(changeRaw.Data, &a)
+				if err != nil {
+					return added, errors.Trace(err)
+				}
+				j, err := bson.MarshalJSON(a)
+				if err != nil {
+					return added, errors.Trace(err)
+				}
+				w.logger.Criticalf("update %s %v without revno %s", change.Ns.Collection, change.DocumentKey.Id, j)
```

Comment:
yep it is, you are watching a collection and a mutation was made outside a sstxn, this is a violation because we require it for correctness. If you are modifying a collection outside a sstxn, it should not be watched.

Reply:
added a comment to help people who get here.